### PR TITLE
basecamps: add modular basecamps

### PR DESCRIPTION
**Pull request by mlangsdorf**
#### Summary
```SUMMARY Features "basecamps: add modular basecamps"```

#### Purpose of change
Addresses #26341

#### Describe the solution
reorganize the basecamp recipe files by moving them into their own file and giving the primitive field camp file a more meaningful name.

add support for basecamps being able to autocalculate their upgrade build time and requirements.  For that to happen, all common terrain constructions need to have a construction definition, but some constructions, like deep and shallow pits, are not meant to be selectable by the user.  add the concept of a hidden construction, and add definitions for the shallow pit and deep pit construction.

add a new "blueprint_autocalculate" field to blueprint recipes.  When it is present, recipe::finalize calls a function that:
1. calls a mapgen_update function that creates a tinymap of dirt, applies the blueprint mapgen_update to that tinymap, and then returns maps of all the non-dirt ter_ids and furn_ids to their respective counts.
2. then passes those maps to function in construction that takes all those ter_ids/furn_ids, finds the construction(s) that result in those ids, and then returns a map of requirement_ids to counts of those requirements.
3. those requirements are then added to the reqs_internal and resolved normally.
4. one more function then eliminates any duplicated tools or qualities and consolidates similar components.

add support for blueprints that don't have an oter_id attached to them, but are pure mapgen_update functions.

also, add a recipe_dictionary entry for blueprints, similar to the existing autolearn entry.

add a new blueprint keyword "blueprint_excludes" which can be used to create mutually exclusive branches in an camp upgrade path.

the list of completed blueprint_provides is compared to an upgrade's blueprint_excludes.  If there is a match, the recipe isn't available.

There is a second list of in_progress provides, which is used to prevent people from building two mutually exclusive upgrades at the same time, while also preventing them from building upgrade B that requires upgrade A before upgrade A is completed.

reorganize the basecamp code slightly by adding a base_camps:: namespace and moving various bits of code into it.

NPCs will work on a single task for up to 11 hours, but on longer tasks, they'll only work 10 hours out of every 24.  Add functions to calculate the workload from the raw task times, and update most of the existing blueprints.

add support for retrieving available basecamp upgrade path starts and expansion by oter_id from the "all_faction_base_types" and "all_faction_base_expansions" recipe groups.  Also populate the expansion tabs in the basecamp control window image the same way.

Let the player select from all available basecamp upgrade path starts if there is more than 1 available for the selected basecamp terrain.

Select available upgrades from recipe_dict's collection of construction blueprints.

add the JSON definitions of the modular field camp.  It can be configured to be very similar to the existing primitive field camp or built as a single large building made from wattle-and-daub, wood panel, or metal.  The player can upgrade it piecemeal in the order they please.  It's also going to be easier to add new features to it in the future.

#### Describe alternatives you've considered
This PR is huge but it includes 3600 lines of JSON changes and additions.  I'd like to break it down into more manageable chunks but it doesn't decompose nicely: there's a bunch of necessary changes up front that doesn't do anything on their own and then suddenly it all comes together.

An early draft of this was submitted in #31055 but I prefer this version.

#### Additional context
Some pretty pictures:

Modular field camp built quickly from tents and wattle-and-daub
MFC_Primitive_complete

Full modular field camp hardshell built from wattle-and-daub
MFC_Semiprimitive_complate

Modular field camp buildings built from a mixture of tents, wattle-and-daub, wood panel, and metal.
MFC_Mix_andMatch

### CODE REVIEW

**Comment by Night-Pryanik** [comment]
> Addresses #26431

I believe that's wrong issue number.

File: doc/BASECAMP.md [hunk 1]
```diff
@@ -121,3 +151,35 @@ or `"faction_base_camp_19"` - building the camp's radio tower, which can be foll
 7.  `"faction_base_camp_17"` - adding better doors to camp wall and the central building must be built after `"faction_base_camp_14"` and `"faction_base_camp_16"`.
 
 After setting up the first tent, the player has a lot of options: they can build additional tents to enable expansions, they can build the well for water, they can build as much or as little of the central kitchen as they desire, or add the wall to give the camp defenses, or the radio tower to improve the range of their two-way radios.
+
+## Modular Basecamp conventions
+The modular basecamp is a structure for designing basecamp upgrade paths.  You don't have to use it, but elements of the design might get more code support in the future.
+
+### Layout
+A modular camp is laid out on a 24x24 overmap tile.  The outer 3 map squares on each side are reserved for fortifications and movement corridors, and the inner 18x18 map squares are divided into a 3x3 grip of 6x6 areas.
```
Comment (davidpwbrown) [inline doc/BASECAMP.md]:
```suggestion
A modular camp is laid out on a 24x24 overmap tile.  The outer 3 map squares on each side are reserved for fortifications and movement corridors, and the inner 18x18 map squares are divided into a 3x3 grid of 6x6 areas.
```

File: src/construction.cpp [hunk 2]
```diff
@@ -1486,3 +1488,83 @@ void finalize_constructions()
...
 }
+
+
+void get_build_reqs_for_furn_ter_ids( const std::pair<std::map<ter_id, int>,
+                                      std::map<furn_id, int>> &changed_ids,
+                                      build_reqs &total_reqs )
+{
+    std::map<size_t, int> total_builds;
+
+    // interatively recurse through the pre-terrains until the pre-terrain is empty, adding
```
Comment (davidpwbrown) [inline src/construction.cpp]:
```suggestion
    // Iteratively recurse through the pre-terrains until the pre-terrain is empty, adding
```

File: src/faction_camp.cpp [hunk 3]
```diff
@@ -434,46 +435,78 @@ static cata::optional<basecamp *> get_basecamp( npc &p )
...
-            om_near.first.find( "river_" ) == std::string::npos ) {
-            popup( _( "You need more room for camp expansions!" ) );
-            return;
+        const oter_id &om_type = oter_id( om_near.first );
+        if( is_ot_subtype( "field", om_type ) ) {
+            near_fields += 1;
         }
     }
+    if( near_fields < 4 ) {
+        popup( _( "You need more at least 4 adjacent  for camp expansions!" ) );
```
Comment (davidpwbrown) [inline src/faction_camp.cpp]: This is intended to specify the terrain needed nearby?

File: src/faction_camp.cpp [hunk 4]
```diff
@@ -434,46 +435,78 @@ static cata::optional<basecamp *> get_basecamp( npc &p )
         return cata::nullopt;
     }
     basecamp *temp_camp = *bcp;
-    temp_camp->define_camp( p );
+    temp_camp->define_camp( p, camp_type );
     return temp_camp;
 }
 
+std::string select_camp_option( const std::map<std::string, std::string> &pos_options,
```
Comment (ZhilkinSerg) [inline src/faction_camp.cpp]:
```suggestion
static std::string select_camp_option( const std::map<std::string, std::string> &pos_options,
```

File: src/faction_camp.cpp [hunk 5]
```diff
@@ -434,46 +435,78 @@ static cata::optional<basecamp *> get_basecamp( npc &p )
...
+                                const std::string &option )
```
Comment (ZhilkinSerg) [inline src/faction_camp.cpp]:
```suggestion
                                    const std::string &option )
```

**Comment by Regularitee** [comment]
Can anyone else confirm/deny this change is the one causing the overmap memory erasing bug on game load?

**Comment by kevingranade** [comment]
It is, Mark has identified the problem and is working on a fix.

**Pull request merged**